gpio/imx6ull-gpio/imx6ull-gpio.c

```diff
@@ -212,13 +208,35 @@ void thread(void *arg)
 				break;
 
 			case mtWrite:
-				if (msg.i.data != NULL && msg.i.size >= sizeof(uint32_t)) {
-					memcpy(&val, msg.i.data, sizeof(uint32_t));
+				if (msg.i.data != NULL && msg.i.size >= 2) {
+					uint8_t *data = (uint8_t *)msg.i.data;
+
+					/* NOTE: ignoring msg.i.io.offs to allow repetetive writes without seek / reopen */
+
+					if ((msg.i.size % sizeof(uint32_t) != 0) && ((data[0] == '+') || (data[0] == '-'))) {
+						printf("GPIO: text mode!\n");
+						/* text mode [+/-][pin] -> 1/0 as out/in or value */
+						uint32_t pin = 0;
+						for (d = 1; d < msg.i.size && isdigit(data[d]); ++d)
+							pin = pin * 10 + (data[d] - '0');
+
+						if (pin > 31)
+							break; /* -EINVAL */
 
-					if (msg.i.size >= sizeof(uint32_t) << 1)
-						memcpy(&mask, msg.i.data + sizeof(uint32_t), sizeof(uint32_t));
-					else
-						mask = (uint32_t)-1;
+						mask = 1 << pin;
+						val = ((data[0] == '+') ? 1 : 0) << pin;
```

Comment:
```suggestion
						mask = 1 << pin;
						val = (data[0] == '+') ? mask : 0;
```
Litle simpler.

Reply:
I prefer my version as it's more semantic (the intent is clearer). With optimizations enabled gcc would produce almost the same code for both versions.